### PR TITLE
Add Item Rarity API

### LINKED ISSUE #5351: Get rarity and proper name from ItemStack and Material
**Is your feature request related to a problem? Please describe.**
It is currently impossible to get the rarity (enum) and rarity color of an ItemStack, other than using NMS reflection. It is also complicated to get the proper name (i.e. "Block of Iron" for `Material.IRON_BLOCK`) for ItemStack or enum Material.

**Describe the solution you'd like**
I would love to see two new functions, for `ItemStack` and `Material`:
- `ItemStack#getRarity`: returns an `enum ItemRarity` (new enum that has to be created, example below)
- `Material#getProperName` (or for `ItemStack`): returns a `String` with the name that is displayed in inventories

The `ItemRarity` would be a new enum, something like this:
```java
public enum ItemRarity
{
    EPIC (ChatColor.LIGHT_PURPLE),
    RARE (ChatColor.AQUA),
    UNCOMMON (ChatColor.YELLOW),
    COMMON (ChatColor.WHITE);

    private final ChatColor chatColor;

    ItemRarity (ChatColor chatColor)
    {
        this.chatColor = chatColor;
    }

    public ChatColor getChatColor ()
    {
        return chatColor;
    }
}
```
Note how the item rarity also already can provide the color that's bound to that rarity.

An example of how I'd like to result to be like:
```java
ItemStack item = new ItemStack(Material.IRON_BLOCK);

String name = item.getType().getProperName(); // "Block of Iron"
// or really can also be `item.getProperName()` if preferred
ItemRarity rarity = item.getRarity(); // ItemRarity.COMMON

System.out.printf("%s, %s, %s", name, rarity.name(), rarity.getChatColor().name());
```
Output: `Block of Iron, COMMON, WHITE`

**Describe alternatives you've considered**
Well, it may or may not be possible to implement it on the `Material` enum instead of the `ItemStack` class, if you can get that done. Not sure. But it would mean it's easier to access, as `Material` is more present (think of `block.getType()`), but idm if it's on `ItemStack`. Though, `Material#getProperName` would be possible, most likely, see examples below.

**Additional context**
Get NMS ItemStack and CraftItemStack for further reference in 1.16.5:
```java
org.bukkit.inventory.ItemStack itemStack = new ItemStack(Material.IRON_BLOCK);
net.minecraft.server.v1_16_R3.Item nmsItem = org.bukkit.craftbukkit.v1_16_R3.util.CraftMagicNumbers.getItem(itemStack.getType());
net.minecraft.server.v1_16_R3.ItemStack craftItemStack = org.bukkit.craftbukkit.v1_16_R3.inventory.CraftItemStack.asNMSCopy(itemStack);
```
How rarity of an `ItemStack` is obtained using NMS, in 1.16.5:
```java
net.minecraft.server.v1_16_R3.EnumItemRarity itemRarity = nmsItem.i(craftItemStack);
```
How the "proper" name of a `Material` is obtained:
```java
String properName = net.md_5.bungee.chat.TranslationRegistry.INSTANCE.translate(nmsItem.getName());
```

### PR DESCRIPTION
**Pull request by Machine-Maker**
Was part of #5351

### CODE REVIEW

File: Spigot-API-Patches/0282-Item-Rarity-API.patch [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
++import net.kyori.adventure.text.format.NamedTextColor;
++import net.kyori.adventure.text.format.TextColor;
++import org.jetbrains.annotations.NotNull;
++
++public enum ItemRarity {
++
++    COMMON(NamedTextColor.WHITE),
++    UNCOMMON(NamedTextColor.YELLOW),
++    RARE(NamedTextColor.AQUA),
++    EPIT(NamedTextColor.LIGHT_PURPLE);
```
Comment (FranKusmiruk) [inline Spigot-API-Patches/0282-Item-Rarity-API.patch]: typo?

Comment (Camotoy) [inline Spigot-API-Patches/0282-Item-Rarity-API.patch]: EPIT?

Comment (Machine-Maker) [reply Spigot-API-Patches/0282-Item-Rarity-API.patch]: man, you guys did that at almost the *exact* same time :laughing:

File: Spigot-Server-Patches/0695-Item-Rarity-API.patch [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
++++ b/src/main/java/org/bukkit/craftbukkit/util/CraftMagicNumbers.java
+@@ -447,6 +447,20 @@ public final class CraftMagicNumbers implements UnsafeValues {
+         return net.minecraft.server.Entity.nextEntityId();
+     }
+ 
++    @Override
++    public io.papermc.paper.inventory.ItemRarity getItemRarity(org.bukkit.Material material) {
++        Item item = getItem(material);
++        if (item == null) {
++            throw new IllegalArgumentException("Only items have rarity");
```
Comment (kashike) [inline Spigot-Server-Patches/0695-Item-Rarity-API.patch]:
```suggestion
+            throw new IllegalArgumentException(material + " is not an item, and rarity does not apply to blocks");
```

**Review by kashike — CHANGES_REQUESTED** [review]

File: Spigot-Server-Patches/0695-Item-Rarity-API.patch (hunk 2 above)
Comment (Machine-Maker) [reply Spigot-Server-Patches/0695-Item-Rarity-API.patch]: yep

**Review by kashike — APPROVED** [review]

**Comment by justEli** [comment]
Now that i think of it, items can have enchants which makes the item have another rarity. i see that it's implemented on Material, but that doesn't take into account enchants. So.. maybe an option to also add support for ItemStack?

**Comment by Machine-Maker** [comment]
I did add smth for ItemStacks on the Unsafe interface.

**Comment by justEli** [comment]
which is.....deprecated, not something i'd likely use

**Pull request merged**